amplipi/audiodetector/audiodetector.c

```diff
@@ -0,0 +1,354 @@
+    uint8_t ch1l : 1;
+    uint8_t ch1r : 1;
+    uint8_t ch2l : 1;
+    uint8_t ch2r : 1;
+    uint8_t ch3l : 1;
+    uint8_t ch3r : 1;
+  };
+  uint8_t all;
+} AudioStatus;
+
+void sigint_handler(int sig) {
+  abort_ = true;
+}
+
+bool register_sig_handler() {
+  struct sigaction sa;
+  memset(&sa, 0, sizeof(sa));
+  sa.sa_handler = sigint_handler;
+  if (sigaction(SIGINT, &sa, NULL) < 0) {
+    perror("sigaction(SIGINT)");
+    return false;
+  }
+  // TODO: What all should be handled?
+  if (sigaction(SIGTERM, &sa, NULL) < 0) {
+    perror("sigaction(SIGTERM)");
+    return false;
+  }
+  return true;
+}
+
+bool open_storage(const char *path, int *fd) {
+  // TODO: O_NOATIME?
+  *fd = open(path, O_RDWR | O_CREAT | O_DSYNC, 0644);
+  if (*fd < 0) {
+    perror("open(file)");
+    return false;
+  }
+  return true;
+}
+
+bool write_data(int fd, AudioStatus *status) {
+  static AudioStatus status_previous = {.all = 0};
+  // Only write data on change
+  if (status->all != status_previous.all) {
+    lseek(fd, 0, SEEK_SET);
+    if (write(fd, status, 1) != 1) {
+      perror("write");
+      return false;
+    }
+    status_previous.all = status->all;
+  }
+  return true;
+}
+
+bool close_storage(const char *path, int fd) {
+  int result = close(fd);
+  if (result < 0) {
+    perror("close(file)");
+    return false;
+  }
+  // Delete output file so AmpliPi knows no status is available.
+  result = remove(path);
+  if (result < 0) {
+    perror("remove(file)");
+    return false;
+  }
+  return true;
+}
+
+bool measure(AdcData *data) {
+  struct spi_ioc_transfer tr[NUM_CHANNELS] = {};
+  uint32_t                tx[NUM_CHANNELS] = {};
+  uint32_t                rx[NUM_CHANNELS] = {};
+
+  for (size_t i = 0; i < NUM_CHANNELS; i++) {
+    tx[i]           = 0x008001 | (i << 12);  // Sends 0x01, 0x80, 0x00
+    tr[i].tx_buf    = (uintptr_t)&tx[i];
+    tr[i].rx_buf    = (uintptr_t)&rx[i];
+    tr[i].len       = 3;
+    tr[i].speed_hz  = SPI_CLOCK_HZ;
+    tr[i].cs_change = 1;
+  }
+
+  // Unset cs_change for the last transfer or we lose the first read of the next
+  // block.
+  tr[NUM_CHANNELS - 1].cs_change = 0;
+
+  int spi_fd = open(SPIDEV_PATH, O_RDWR);
+  if (spi_fd < 0) {
+    perror("open(spi)");
+    memset(data, 0, sizeof(AdcData));
+    return false;
+  }
+
+  /* This is the default, but leaving here as reference in case that changes.
+   * uint32_t mode = SPI_MODE_0;
+   * ioctl(spi_fd, SPI_IOC_WR_MODE, &mode);
+   */
+
+  if (ioctl(spi_fd, SPI_IOC_MESSAGE(NUM_CHANNELS), tr) < 0) {
+    perror("ioctl()");
+  }
+
+  // ADC channel 7 = RCA 0 Right, ADC channel 6 = RCA 0 Left
+  // ADC channel 5 = RCA 1 Right, etc...
+  for (size_t i = 0; i < NUM_CHANNELS; i++) {
+    data->vals[i] = (rx[i] & 0x300) + ((rx[i] >> 16) & 0xFF);  // rx[i] >> 8;
+  }
+
+  close(spi_fd);
+  return true;
+}
+
+void update_window(AdcWindow *window, AdcData *raw_data) {
+  // window is implemented like a circular queue
+  static size_t window_index = 0;
```

Comment:
no one else needs window_index since we are treating each time step in the window indiscriminately (just taking the max across it). its only used on write. do you want it bundled with AdcWindow for the sake of separating state from logic?

Reply:
yeah, it just feels weird having coupled state in 2 places like that.